stdlib/public/TensorFlow/CompilerRuntime.swift

```diff
@@ -394,6 +579,154 @@ public final class _ExecutionContext {
+      let newHandle = TFE_TensorHandleCopySharingTensor(address.pointee,
+                                                        status)
+      checkOk(status)
+      debugLog("""
+                 Copying the \(idx)-th C handle \(address.pointee) with \
+                 concrete=\(isConcrete) to C handle create \(newHandle!).
+                 """)
+      output.append(newHandle!)
+    }
+    TF_DeleteStatus(status)
+    return output
+  }
+}
+
+public extension TensorGroup {
+  // The tensors in `input` are NOT owned by this instance.
+  init(_copying input: [CTensorHandle]) {
+    assert(Self._tensorHandleCount == input.count)
+    let buffer = UnsafeMutablePointer<CTensorHandle>.allocate(capacity: input.count)
+    let status = TF_NewStatus()
+    // copy input to buffer
+    for (idx, inputTensorHandle) in input.enumerated() {
+      let address = buffer.advanced(by: idx)
+      // Each tensor can be symbolic (e.g. when using this API to create a
+      // symbolic input instance to tracee) or concrete (e.g. when creating the
+      // final output of the tracee).
+      let newHandle = TFE_TensorHandleCopySharingTensor(inputTensorHandle, status)
+      checkOk(status)
+      address.initialize(to: newHandle!)
+    }
+    TF_DeleteStatus(status)
+    self.init(_owning: buffer)
+  }
+}
+
+// TODO: assess if this protocol should be folded into TensorArrayProtocol.
+public protocol TensorArrayProtocolEnhanced : TensorArrayProtocol {
+  // Create an instance based on `inputs`, which can be symbolic (e.g. when
+  // creating a symbolic input to tracee) or concrete (e.g. when creating a
+  // final output of executing the tracee).
+  func createInstance(_owning inputs: [CTensorHandle]) -> Self
+}
+
+public func trace<State : TensorArrayProtocolEnhanced,
+                  Data : TensorGroup,
+                  Result : TensorGroup>(
+  with state: State,
+  in fn: (State, Data) -> (State, Result)
+) -> (State, Data) -> (State, Result) {
+  debugLog("""
+             Tracing over a function with \(state._tensorHandleCount) input \
+             state tensors and \(Data._typeList.count) input data tensors.
+             """)
+
+  // Verify that we are not already tracing.
+  internalConsistencyCheck(_RuntimeConfig.traceState.getFunction == nil,
+                           "Should not be in tracing mode already!")
+
+  // Switch to tracing mode.
+  let inputValueCount = Int(state._tensorHandleCount) + Data._typeList.count
+  _RuntimeConfig.traceState = .tracing(
+    TraceContext(inputValueCount: inputValueCount))
+
+  // Handle inputs.
+  let traceFn = _RuntimeConfig.traceState.getFunction!
+  // TODO: support other dtypes.
+  let inputSymbolicTensors = traceFn.symbolicInputs.map {
+    TFE_NewTensorHandleFromTFOutput($0, TF_FLOAT)!
+  }
+  internalConsistencyCheck(inputSymbolicTensors.count == inputValueCount)
+  let symbolicState = state.createInstance(
+    _owning: Array(inputSymbolicTensors.dropLast(Data._typeList.count)))
```

Comment:
It's better if you make this method (`TensorArrayProtocol._makeInstance(_owning:)` or `TensorArrayProtocol.init(_owning:)`) take a generic type `C : Collection where C.Element == CTensorHandle`. Otherwise it's inefficient (you had to create a new `Array` every time) and complicates callers.

Reply:
Nice suggestion! Done.